src/util/bson/bson.hpp

```diff
@@ -0,0 +1,190 @@
+#include "regular_expression.hpp"
+#include "min_key.hpp"
+#include "max_key.hpp"
+#include "null.hpp"
+
+#include <realm/binary_data.hpp>
+#include <realm/timestamp.hpp>
+#include <realm/decimal128.hpp>
+#include <realm/object_id.hpp>
+#include <iostream>
+#include <unordered_map>
+#include <variant>
+
+namespace realm {
+namespace bson {
+
+class Bson;
+
+/// A variant of the allowed Bson types.
+class Bson : public std::variant<
+    Null,
+    int32_t,
+    int64_t,
+    bool,
+    float,
+    double,
+    time_t,
+    std::string,
+    std::vector<char>,
+    Timestamp,
+    Decimal128,
+    ObjectId,
+    RegularExpression,
+    MinKey,
+    MaxKey,
+    IndexedMap<Bson>,
+    std::vector<Bson>
+> {
+public:
+    using base = std::variant<
+        Null,
+        int32_t,
+        int64_t,
+        bool,
+        float,
+        double,
+        time_t,
+        std::string,
+        std::vector<char>,
+        Timestamp,
+        Decimal128,
+        ObjectId,
+        RegularExpression,
+        MinKey,
+        MaxKey,
+        IndexedMap<Bson>,
+        std::vector<Bson>
+    >;
+    using base::base;
+    using base::operator=;
+};
+
+using BsonDocument = IndexedMap<Bson>;
+using BsonArray = std::vector<Bson>;
+
+template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+
+inline std::ostream& operator<<(std::ostream& out, const Bson& b)
+{
+    std::visit(overloaded {
+        [&out](auto) {
+            out << "null";
+        },
+        [&out](Null) {
+            out << "null";
+        },
+        [&out](int32_t bson) {
+            out << "{" << "\"$numberInt\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&out](double bson) {
+            out << "{" << "\"$numberDouble\"" << ":" << '"';
+            if (isnan(bson)) {
+                out << "NaN";
+            } else if (bson == std::numeric_limits<double>::infinity()) {
+                out << "Infinity";
+            } else if (bson == (-1 * std::numeric_limits<double>::infinity())) {
+                out << "-Infinity";
+            } else {
+                out << bson;
+            }
+            out << '"' << "}";
+        },
+        [&out](int64_t bson) {
+            out << "{" << "\"$numberLong\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&out](Decimal128 bson) {
+            out << "{" << "\"$numberDecimal\"" << ":" << '"';
+            if (bson.is_nan()) {
+                out << "NaN";
+            } else if (bson == Decimal128("Infinity")) {
+                out << "Infinity";
+            } else if (bson == Decimal128("-Infinity")) {
+                out << "-Infinity";
+            } else {
+                out << bson;
+            }
+            out << '"' << "}";
+        },
+        [&out](ObjectId bson) {
+            out << "{" << "\"$oid\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&out](BsonDocument bson) {
```

Comment:
Copying the argument for BsonDocument and BsonArray seems like it could easily matter sometimes.